File: spec/wuzzy-nest/create-crawler.spec.lua
local codepath = 'wuzzy-nest.wuzzy-nest'

describe('Wuzzy-Nest Create-Crawler', function()
  local WuzzyNest = require(codepath)
  before_each(function()
    CacheOriginalGlobals()
    WuzzyNest = require(codepath)
  end)
  after_each(function()
    RestoreOriginalGlobals()
    package.loaded[codepath] = nil
  end)

  it('spawns a new Wuzzy-Crawler process & tracks the spawn', function()
    _G.send = spy.new(function() end)
    _G.spawn = spy.new(function() end)
    local handler = GetHandler('Create-Crawler')
    local msgId = 'mock-message-id-1'
    local from = _G.owner
    local crawlerName = 'My Wuzzy Crawler'

    handler.handle({
      id = msgId,
      from = from,
      target = 'wuzzy-nest-process-id',
      action = 'Create-Crawler'
    })

    assert.spy(_G.spawn).was.called_with(_G.module, {
      ['App-Name'] = 'Wuzzy',
      ['Contract-Name'] = 'wuzzy-crawler',
      authority = _G.authorities[1],
      ['X-Create-Crawler-Id'] = msgId,
      ['Crawler-Name'] = crawlerName,
      ['Crawler-Creator'] = from
    })
    assert.are_same(WuzzyNest.State.CrawlerSpawnRefs[msgId], {
      Creator = from,
      CrawlerName = crawlerName
    })
    assert.spy(_G.send).was.called_with({
      target = from,
      action = 'Create-Crawler-Result',
      data = 'OK',
      ['X-Create-Crawler-Id'] = msgId
    })
  end)

  it('notifies creator of new crawler spawn & tracks creators', function()
    _G.send = spy.new(function() end)
    _G.spawn = spy.new(function() end)
    local msgId = 'mock-message-id-1'
    local from = _G.owner
    local newCrawlerProcessId = 'new-crawler-process-id'
    local crawlerName = 'My Wuzzy Crawler'

    GetHandler('Create-Crawler').handle({
      id = msgId,
      from = from,
      target = _G.id,
      action = 'Create-Crawler'
    })
    _G.send:clear()
    _G.spawn:clear()
    GetHandler('Spawned').handle({
      id = 'mock-message-id-2',
      from = _G.id,
      target = _G.id,
      action = 'Spawned',
      ['X-Create-Crawler-Id'] = msgId,
      ['From-Process'] = _G.id,
      ['Process'] = newCrawlerProcessId
    })

    assert.spy(_G.send).was.called_with({
      target = newCrawlerProcessId,
      action = 'Eval',
      data = require('.base64').decode(
        require('..wuzzy-crawler.wuzzy-crawler-stringified')
      )
    })
    assert.are_same(WuzzyNest.State.Crawlers[newCrawlerProcessId], {
      ['X-Create-Crawler-Id'] = msgId,
      Creator = from,
      Name = crawlerName,
      Owner = from,
      Roles = {}
    })
    assert.is_nil(WuzzyNest.State.CrawlerSpawnRefs[msgId])
    assert.spy(_G.send).was.called_with({
      target = from,
      action = 'Crawler-Spawned',
      ['Crawler-Id'] = newCrawlerProcessId,
      ['X-Create-Crawler-Id'] = msgId,
      data = 'OK'
    })
  end)

  it('throws on Create-Crawler messages from unauthorized callers', function()
    assert.has.errors(function()
      GetHandler('Create-Crawler').handle({
        id = 'mock-message-id-1',
        from = 'alice-mock-address',
        target = 'wuzzy-nest-process-id',
        action = 'Create-Crawler'
      })
    end, 'Permission Denied')
  end)

  it('ignores Spawned messages without known X-Create-Crawler-Id', function()
    _G.send = spy.new(function() end)

    GetHandler('Spawned').handle({
      id = 'mock-message-id-2',
      from = _G.id,
      target = _G.id,
      action = 'Spawned',
      ['From-Process'] = _G.id,
      ['Process'] = 'new-crawler-process-id'
    })

    assert.spy(_G.send).was.called(0)
  end)

  it('ignores Spawned messages from unknown process', function()
    _G.send = spy.new(function() end)

    GetHandler('Spawned').handle({
      id = 'mock-message-id-2',
      from = _G.id,
      target = _G.id,
      action = 'Spawned',
      ['X-Create-Crawler-Id'] = 'unknown-id',
      ['From-Process'] = _G.id,
      ['Process'] = 'new-crawler-process-id'
    })

    assert.spy(_G.send).was.called(0)
  end)

  it('ignores Spawned messages from unknown caller', function()
    _G.send = spy.new(function() end)

    GetHandler('Spawned').handle({
      id = 'mock-message-id-2',
      from = 'unknown-caller-id',
      target = _G.id,
      action = 'Spawned',
      ['X-Create-Crawler-Id'] = 'unknown-id',
      ['From-Process'] = _G.id,
      ['Process'] = 'new-crawler-process-id'
    })

    assert.spy(_G.send).was.called(0)
  end)

  it('ignores Spawned messages with unknown X-Create-Crawler-Id', function()
    _G.send = spy.new(function() end)

    GetHandler('Spawned').handle({
      id = 'mock-message-id-2',
      from = _G.id,
      target = _G.id,
      action = 'Spawned',
      ['X-Create-Crawler-Id'] = 'unknown-id',
      ['From-Process'] = _G.id,
      ['Process'] = 'new-crawler-process-id'
    })

    assert.spy(_G.send).was.called(0)
  end)

  it('ignores Spawned messages without Process tag', function()
    _G.send = spy.new(function() end)
    local xCreateCrawlerId = 'mock-create-crawler-id'

    WuzzyNest.State.CrawlerSpawnRefs[xCreateCrawlerId] = {
      Creator = _G.owner,
      CrawlerName = 'My Wuzzy Crawler'
    }

    GetHandler('Spawned').handle({
      id = 'mock-message-id-2',
      from = _G.id,
      target = _G.id,
      action = 'Spawned',
      ['X-Create-Crawler-Id'] = xCreateCrawlerId,
      ['From-Process'] = _G.id,
    })

    assert.spy(_G.send).was.called(0)
  end)

  pending('uses ~patch@1.0 whenever updating state', function() end)
end)

File: test/spec/contracts/wuzzy-nest/create-crawler.spec.ts
import { expect } from 'chai'

import {
  ALICE_ADDRESS,
  AOTestHandle,
  ARIO_NETWORK_PROCESS_ID,
  AUTHORITY_ADDRESS,
  createLoader,
  MODULE_ID,
  ORACLE_ADDRESS,
  OWNER_ADDRESS,
  PROCESS_ID
} from '~/test/util/setup'

describe('Wuzzy-Nest Create-Crawler', () => {
  let handle: AOTestHandle

  beforeEach(async () => {
    handle = (await createLoader('wuzzy-nest')).handle
  })

  it('spawns a new Wuzzy-Crawler process & tracks the spawn', async () => {
      const result = await handle({
        From: OWNER_ADDRESS,
        Tags: [ { name: 'Action', value: 'Create-Crawler' } ]
      })
      expect(result.Error).to.not.exist
      expect(result.Spawns).to.have.lengthOf(1)
      expect(result.Spawns[0].Tags).to.deep.include({
        name: 'Module',
        value: MODULE_ID
      })
      expect(result.Spawns[0].Tags).to.deep.include({
        name: 'Contract-Name',
        value: 'wuzzy-crawler'
      })
      expect(result.Spawns[0].Tags).to.deep.include({
        name: 'App-Name',
        value: 'Wuzzy'
      })
      expect(result.Spawns[0].Tags).to.deep.include({
        name: 'Authority',
        value: AUTHORITY_ADDRESS
      })
      const spawnRef = result.Spawns[0].Tags.find(
        tag => tag.name === 'X-Create-Crawler-Id'
      )?.value
      expect(spawnRef).to.exist
  
      expect(result.Messages).to.have.lengthOf(1)
      expect(result.Messages[0].Data).to.exist
      expect(result.Messages[0].Data).to.equal('OK')
      expect(result.Messages[0].Tags).to.deep.include({
        name: 'Action',
        value: 'Create-Crawler-Result'
      })
      expect(result.Messages[0].Tags).to.deep.include({
        name: 'X-Create-Crawler-Id',
        value: spawnRef
      })
  
      const viewStateResult = await handle({
        From: OWNER_ADDRESS,
        Tags: [ { name: 'Action', value: 'View-State' } ]
      })
      expect(viewStateResult.Error).to.not.exist
      expect(viewStateResult.Messages).to.have.lengthOf(1)
      expect(viewStateResult.Messages[0].Data).to.exist
      const state = JSON.parse(viewStateResult.Messages[0].Data)
      expect(state.CrawlerSpawnRefs[spawnRef!]).to.exist
      expect(state.CrawlerSpawnRefs[spawnRef!]).to.deep.equal({
        Creator: OWNER_ADDRESS,
        CrawlerName: 'My Wuzzy Crawler'
      })
  })

  it('notifies creator of new crawler spawn & tracks crawler creators', async () => {
    const newProcessId = 'new-process-id'.padEnd(43, '0')
    const result = await handle({
      From: OWNER_ADDRESS,
      Tags: [ { name: 'Action', value: 'Create-Crawler' } ]
    })
  
    expect(result.Error).to.not.exist
    const spawnRef = result.Spawns[0].Tags.find(
      tag => tag.name === 'X-Create-Crawler-Id'
    )?.value
    expect(spawnRef).to.exist

    const spawnedResult = await handle({
      From: PROCESS_ID,
      Tags: [
        { name: 'Action', value: 'Spawned' },
        { name: 'X-Create-Crawler-Id', value: spawnRef! },
        { name: 'From-Process', value: PROCESS_ID },
        { name: 'Process', value: newProcessId }
      ]
    })
    expect(spawnedResult.Error).to.not.exist
    expect(spawnedResult.Messages).to.have.lengthOf(2)

    expect(spawnedResult.Messages[0].Target).to.equal(newProcessId)
    expect(spawnedResult.Messages[0].Tags).to.deep.include({
      name: 'Action',
      value: 'Eval'
    })
    expect(spawnedResult.Messages[0].Tags).to.deep.include({
      name: 'Ario-Network-Process-Id',
      value: ARIO_NETWORK_PROCESS_ID
    })
    expect(spawnedResult.Messages[0].Tags).to.deep.include({
      name: 'Data-Oracle-Address',
      value: ORACLE_ADDRESS
    })
    expect(spawnedResult.Messages[0].Data).to.exist

    expect(spawnedResult.Messages[1].Target).to.equal(OWNER_ADDRESS)
    expect(spawnedResult.Messages[1].Tags).to.deep.include({
      name: 'Action',
      value: 'Crawler-Spawned'
    })
    expect(spawnedResult.Messages[1].Data).to.equal('OK')
    expect(spawnedResult.Messages[1].Tags).to.deep.include({
      name: 'Crawler-Id',
      value: newProcessId
    })

    const viewStateResult = await handle({
      From: OWNER_ADDRESS,
      Tags: [ { name: 'Action', value: 'View-State' } ]
    })
    expect(viewStateResult.Error).to.not.exist
    expect(viewStateResult.Messages).to.have.lengthOf(1)
    expect(viewStateResult.Messages[0].Data).to.exist
    const state = JSON.parse(viewStateResult.Messages[0].Data)
    expect(state.Crawlers[newProcessId]).to.exist
    expect(state.Crawlers[newProcessId]).to.deep.equal({
      Ref: spawnRef,
      Name: 'My Wuzzy Crawler',
      Creator: OWNER_ADDRESS,
      Owner: OWNER_ADDRESS,
      Roles: []
    })
  })

  it('throws on Create-Crawler messages from unauthorized callers', async () => {
    const result = await handle({
      From: ALICE_ADDRESS,
      Tags: [ { name: 'Action', value: 'Create-Crawler' } ]
    })
    expect(result.Error).to.exist
    expect(result.Error).to.include('Permission Denied')
  })

  it('ignores Spawned messages without known X-Create-Crawler-Id', async () => {
    const spawnedResult = await handle({
      From: OWNER_ADDRESS,
      Tags: [
        { name: 'Action', value: 'Spawned' },
        { name: 'X-Create-Crawler-Id', value: 'spawn-ref' },
        { name: 'From-Process', value: PROCESS_ID },
        { name: 'Process', value: 'new-process-id'.padEnd(43, '0') }
      ]
    })

    expect(spawnedResult.Error).to.not.exist
    expect(spawnedResult.Messages).to.have.lengthOf(0)
  })

  it('ignores Spawned messages from unknown process', async () => {
    const result = await handle({
      From: OWNER_ADDRESS,
      Tags: [ { name: 'Action', value: 'Create-Crawler' } ]
    })
  
    expect(result.Error).to.not.exist
    expect(result.Spawns).to.have.lengthOf(1)
    const spawnRef = result.Spawns[0].Tags.find(
      tag => tag.name === 'X-Create-Crawler-Id'
    )?.value
    expect(spawnRef).to.exist
    
    const spawnedResult = await handle({
      From: OWNER_ADDRESS,
      Tags: [
        { name: 'Action', value: 'Spawned' },
        { name: 'X-Create-Crawler-Id', value: spawnRef! },
        { name: 'From-Process', value: 'unknown-process'.padEnd(43, 'q') },
        { name: 'Process', value: 'new-process-id'.padEnd(43, '0')}
      ]
    })
    expect(spawnedResult.Error).to.not.exist
    expect(spawnedResult.Messages).to.have.lengthOf(0)
  })

  it('ignores Spawned messages from unknown caller', async () => {
    const newProcessId = 'new-process-id'.padEnd(43, '0')
    const createResult = await handle({
      From: OWNER_ADDRESS,
      Tags: [ { name: 'Action', value: 'Create-Crawler' } ]
    })
    expect(createResult.Error).to.not.exist
    const spawnRef = createResult.Spawns[0].Tags.find(
      tag => tag.name === 'X-Create-Crawler-Id'
    )?.value
    expect(spawnRef).to.exist

    const result = await handle({
      From: OWNER_ADDRESS,
      Tags: [
        { name: 'Action', value: 'Spawned' },
        { name: 'X-Create-Crawler-Id', value: spawnRef! },
        { name: 'From-Process', value: PROCESS_ID },
        { name: 'Process', value: newProcessId }
      ]
    })

    expect(result.Error).to.not.exist
    expect(result.Messages).to.have.lengthOf(0)
  })

  it('ignores Spawned messages without any X-Create-Crawler-Id tag', async () => {
    const result = await handle({
      From: OWNER_ADDRESS,
      Tags: [ { name: 'Action', value: 'Create-Crawler' } ]
    })
    expect(result.Error).to.not.exist

    const spawnedResult = await handle({
      From: PROCESS_ID,
      Tags: [
        { name: 'Action', value: 'Spawned' },
        { name: 'From-Process', value: PROCESS_ID },
        { name: 'Process', value: 'new-process-id'.padEnd(43, '0') }
      ]
    })

    expect(spawnedResult.Error).to.not.exist
    expect(spawnedResult.Messages).to.have.lengthOf(0)
  })

  it('ignores Spawned messages without Process tag', async () => {
    const result = await handle({
      From: OWNER_ADDRESS,
      Tags: [ { name: 'Action', value: 'Create-Crawler' } ]
    })
    expect(result.Error).to.not.exist
    const spawnRef = result.Spawns[0].Tags.find(
      tag => tag.name === 'X-Create-Crawler-Id'
    )?.value
    expect(spawnRef).to.exist

    const spawnedResult = await handle({
      From: OWNER_ADDRESS,
      Tags: [
        { name: 'Action', value: 'Spawned' },
        { name: 'X-Create-Crawler-Id', value: spawnRef! },
        { name: 'From-Process', value: PROCESS_ID }
      ]
    })

    expect(spawnedResult.Error).to.not.exist
    expect(spawnedResult.Messages).to.have.lengthOf(0)
  })

  it('clears spawn refs on Spawned messages', async () => {
    const result = await handle({
      From: OWNER_ADDRESS,
      Tags: [ { name: 'Action', value: 'Create-Crawler' } ]
    })
    expect(result.Error).to.not.exist
    const spawnRef = result.Spawns[0].Tags.find(
      tag => tag.name === 'X-Create-Crawler-Id'
    )?.value
    expect(spawnRef).to.exist
    const spawnedResult = await handle({
      From: PROCESS_ID,
      Tags: [
        { name: 'Action', value: 'Spawned' },
        { name: 'X-Create-Crawler-Id', value: spawnRef! },
        { name: 'From-Process', value: PROCESS_ID },
        { name: 'Process', value: 'new-process-id'.padEnd(43, '0') }
      ]
    })
    expect(spawnedResult.Error).to.not.exist

    const viewStateResult = await handle({
      From: OWNER_ADDRESS,
      Tags: [ { name: 'Action', value: 'View-State' } ]
    })
    expect(viewStateResult.Error).to.not.exist
    expect(viewStateResult.Messages).to.have.lengthOf(1)
    expect(viewStateResult.Messages[0].Data).to.exist
    const state = JSON.parse(viewStateResult.Messages[0].Data)
    expect(state.CrawlerSpawnRefs[spawnRef!]).to.not.exist
  })
})
